frontend/src/main/scala/bloop/bsp/BloopBspServices.scala

```diff
@@ -889,12 +889,21 @@ final class BloopBspServices(
   ): BspEndpointResponse[bsp.WorkspaceBuildTargetsResult] = {
     // need a separate block so so that state is refreshed after regenerating project data
     val refreshTask = ifInitialized(None) { (state: State, logger: BspServerLogger) =>
-      val exitCode = state.client.refreshProjectsCommand
+      state.client.refreshProjectsCommand
         .map { command =>
-          Forker.run(cwd = state.build.origin, command, logger, state.commonOptions)
+          val cwd = state.build.origin
+          Forker.run(cwd = cwd, command, logger, state.commonOptions).map { exitCode =>
+            val status = Forker.exitStatus(exitCode)
+            if (!status.isOk) {
+              val message =
+                s"""Refresh projects command failed with exit code $exitCode.
+                   |Command run in $cwd: ${command.mkString(" ")}""".stripMargin
+              throw new RuntimeException(message)
```

Comment:
I think you can avoid throwing the exception and return a `Left` with the message

Reply:
Right. It wasn't working before because I was actually ignoring this left from refresh and always proceeded with lodaing targets. Now it's ok.